### PR TITLE
Adding support for multiple values for custom fields of type version

### PR DESCRIPTION
**Pull request by Rincewind34**
**Changes I made:**
- Showing the "Allow Multiple Values" check box in the admin interface for custom fields of type version. (same as for user)
- Passing the display field rendering to the same module that renders multiple values of custom fields type "list"

**Manual tests I performed:**
- Being able to create a new version custom field with and without allowing multiple versions.
- Adding the new custom field to a work package form
- Editing the custom field's value at various places: single-wp-view, inline in table, side-wp-view.
- Filter for the custom field with all four supported "multiple value" filters "is", "is not", "all", "none"

Concerning: https://community.openproject.org/projects/openproject/work_packages/40361
Concrete feature: https://community.openproject.org/projects/openproject/work_packages/41060

### CODE REVIEW

**Comment by oliverguenther** [comment]
Hi @Rincewind34, thanks for taking the time and effort to start working on this feature. The changes look good to me thus far! To be able to merge this, we'll need to have added automated tests that extend for the currently single-version fields:

spec/features/custom_fields/custom_fields_spec.rb
spec/features/custom_fields/multi_value_custom_field_spec.rb
spec/features/work_packages/details/custom_fields/custom_field_spec.rb

Best
Oliver

**Comment by Rincewind34** [comment]
Thanks @oliverguenther. I wrote a mail concerning the tests since I don't seem to be able to run them on MacOS.

**Comment by Rincewind34** [comment]
@oliverguenther I'd really love to get stuff done here, but no one is answering my questions... https://community.openproject.org/topics/14552?r=14552#message-14552

**Comment by oliverguenther** [comment]
> @oliverguenther I'd really love to get stuff done here, but no one is answering my questions... https://community.openproject.org/topics/14552?r=14552#message-14552



I added some pointers in the forum, I hope this will help you get started.

Best
Oliver

**Comment by lindenthal** [comment]
Hi @Rincewind34,

how is it going with this PR? Are you planning to finish this?

Best
Niels

**Comment by Rincewind34** [comment]
Hi @lindenthal,
I‘m afraid not. I couldn’t get your tests to run with Mac and even if I‘d do, writing/understanding the tests has proven to be very uneconomic for me. I maintain my own fork of openproject with all the features I need, and will just keep providing my point of view for your work packages I’m interested in :)

**Comment by Rincewind34** [comment]
@oliverguenther Hi Oliver,
since I'm finally able to run the unit tests, I decided to have a quick look at this again. But I really don't know what I am supposed to do. I checked all files and non of them even mention "version". Does that mean that up until now the version custom field type was just left out entirely?

I'm totally willing to write tests for my contributions - I'd just like to understand what and how I am supposed to do that. I read the resource you provided about rspec but it really didn't tell me anything I couldn't infer from just looking at the code.

So bottom line: I need to know how I get started to include the new version array custom field into the existing tests. I need to know which test cases I should write. In theory, since I really just made a tiny change to the backend, a tiny change in tests should be sufficient, if everything else is already tested.

_I'm developing for over 12 years, you don't have go slow on me, just point me in the right direction and I'll get it figured out_ :)

**Comment by oliverguenther** [comment]
Hi @Rincewind34 ,

a good pointer is grepping for `version_.+_custom_field` factory usages. These are work package, project, etc. custom fields of type version. Some of these specs will have to be extended to work with multi select

```
rg -e "version_.+_custom_field"                                                   (feature/45060/reactivate-comment-draft-on-reenter)dev
modules/dashboards/spec/features/project_details_spec.rb
34:  let!(:version_cf) { create(:version_project_custom_field) }

spec/lib/api/v3/work_packages/eager_loading/custom_value_integration_spec.rb
55:      create(:version_wp_custom_field, name: 'blubs').tap do |cf|

spec/lib/api/v3/work_packages/schema/typed_work_package_schema_spec.rb
107:    let(:version_cf) { build_stubbed(:version_wp_custom_field) }

spec/lib/api/v3/work_packages/schema/specific_work_package_schema_spec.rb
321:    let(:version_cf) { build_stubbed(:version_wp_custom_field) }

spec/lib/api/v3/utilities/custom_field_injector_spec.rb
150:        build(:version_wp_custom_field,

spec/lib/api/v3/projects/project_representer_rendering_spec.rb
83:  let(:version_custom_field) { build_stubbed(:version_project_custom_field, visible: true) }

spec/lib/api/v3/queries/schemas/filter_dependency_representer_factory_spec.rb
143:        let(:custom_field) { build_stubbed(:version_wp_custom_field) }

spec/factories/custom_field_factory.rb
57:      factory :version_project_custom_field do
191:      factory :version_wp_custom_field do

spec/models/work_package/work_package_acts_as_customizable_spec.rb
62:      let(:version_cf) { create(:version_wp_custom_field, is_required: true) }

spec/features/projects/projects_custom_fields_spec.rb
43:      create(:version_project_custom_field)

spec/models/custom_actions/actions/custom_field_spec.rb
44:    build_stubbed(:version_wp_custom_field)

spec/models/projects/exporter/exportable_project_context.rb
30:  shared_let(:version_cf) { create(:version_project_custom_field, position: 1) }

spec/models/queries/projects/filters/custom_field_filter_spec.rb
38:  let(:version_project_custom_field) { build_stubbed(:version_project_custom_field) }
49:     version_project_custom_field,
184:      let(:cf_accessor) { "cf_#{version_project_custom_field.id}" }
279:      let(:cf_accessor) { "cf_#{version_project_custom_field.id}" }
283:        allow(version_project_custom_field)
347:      expect(filters.detect { |filter| filter.name == :"cf_#{version_project_custom_field.id}" })
441:      let(:custom_field) { version_project_custom_field }

spec/models/queries/work_packages/filter/custom_fields/custom_field_filter_spec.rb
38:  let(:version_wp_custom_field) { build_stubbed(:version_wp_custom_field) }
49:     version_wp_custom_field,
184:      let(:cf_accessor) { "cf_#{version_wp_custom_field.id}" }
281:      let(:cf_accessor) { "cf_#{version_wp_custom_field.id}" }
285:        allow(version_wp_custom_field)
392:        expect(filters.detect { |filter| filter.name == :"cf_#{version_wp_custom_field.id}" })
488:      let(:custom_field) { version_wp_custom_field }

spec/contracts/projects/shared_contract_examples.rb
235:      let(:custom_field) { build_stubbed(:version_project_custom_field) }
oliver@Olivers-MBP dev % rg -e "version_.+_custom_field"                                                   (feature/45060/reactivate-comment-draft-on-reenter)dev
modules/dashboards/spec/features/project_details_spec.rb
34:  let!(:version_cf) { create(:version_project_custom_field) }

spec/models/custom_actions/actions/custom_field_spec.rb
44:    build_stubbed(:version_wp_custom_field)

spec/models/queries/projects/filters/custom_field_filter_spec.rb
38:  let(:version_project_custom_field) { build_stubbed(:version_project_custom_field) }
49:     version_project_custom_field,
184:      let(:cf_accessor) { "cf_#{version_project_custom_field.id}" }
279:      let(:cf_accessor) { "cf_#{version_project_custom_field.id}" }
283:        allow(version_project_custom_field)
347:      expect(filters.detect { |filter| filter.name == :"cf_#{version_project_custom_field.id}" })
441:      let(:custom_field) { version_project_custom_field }

spec/models/queries/work_packages/filter/custom_fields/custom_field_filter_spec.rb
38:  let(:version_wp_custom_field) { build_stubbed(:version_wp_custom_field) }
49:     version_wp_custom_field,
184:      let(:cf_accessor) { "cf_#{version_wp_custom_field.id}" }
281:      let(:cf_accessor) { "cf_#{version_wp_custom_field.id}" }
285:        allow(version_wp_custom_field)
392:        expect(filters.detect { |filter| filter.name == :"cf_#{version_wp_custom_field.id}" })
488:      let(:custom_field) { version_wp_custom_field }

spec/models/projects/exporter/exportable_project_context.rb
30:  shared_let(:version_cf) { create(:version_project_custom_field, position: 1) }

spec/models/work_package/work_package_acts_as_customizable_spec.rb
62:      let(:version_cf) { create(:version_wp_custom_field, is_required: true) }

spec/contracts/projects/shared_contract_examples.rb
235:      let(:custom_field) { build_stubbed(:version_project_custom_field) }

spec/features/projects/projects_custom_fields_spec.rb
43:      create(:version_project_custom_field)

spec/factories/custom_field_factory.rb
57:      factory :version_project_custom_field do
191:      factory :version_wp_custom_field do

spec/lib/api/v3/work_packages/schema/typed_work_package_schema_spec.rb
107:    let(:version_cf) { build_stubbed(:version_wp_custom_field) }

spec/lib/api/v3/work_packages/eager_loading/custom_value_integration_spec.rb
55:      create(:version_wp_custom_field, name: 'blubs').tap do |cf|

spec/lib/api/v3/work_packages/schema/specific_work_package_schema_spec.rb
321:    let(:version_cf) { build_stubbed(:version_wp_custom_field) }

spec/lib/api/v3/projects/project_representer_rendering_spec.rb
83:  let(:version_custom_field) { build_stubbed(:version_project_custom_field, visible: true) }

spec/lib/api/v3/queries/schemas/filter_dependency_representer_factory_spec.rb
143:        let(:custom_field) { build_stubbed(:version_wp_custom_field) }

spec/lib/api/v3/utilities/custom_field_injector_spec.rb
150:        build(:version_wp_custom_field,
```


and then for pointers on how to test multi-select, grep for existing factories that use `multi_value: true`

```
% r 'multi_value: true'                                                            (feature/45060/reactivate-comment-draft-on-reenter)dev
services/projects/copy_service_integration_spec.rb
728:        let(:list_custom_field) { create(:list_project_custom_field, multi_value: true) }

lib/journal_formatter/custom_field_spec.rb
246:      build_stubbed(:list_wp_custom_field, multi_value: true).tap do |cf|

models/query/results_spec.rb
183:               multi_value: true).tap do |cf|

models/custom_actions/actions/custom_field_spec.rb
41:                  multi_value: true)

features/users/invite_user_modal/custom_fields_spec.rb
55:  let!(:list_multi_cf) { create :list_user_custom_field, name: 'Multi list', multi_value: true, is_required: true }

models/custom_value/list_strategy_integration_spec.rb
38:      multi_value: true,

features/custom_fields/multi_user_custom_field_spec.rb
22:      multi_value: true,

models/work_package/work_package_multi_value_custom_fields_spec.rb
39:      multi_value: true,

features/projects/create_spec.rb
67:    let!(:list_custom_field) { create(:list_project_custom_field, name: 'List CF', multi_value: true) }

features/work_packages/timeline/timeline_labels_spec.rb
69:      multi_value: true,

features/work_packages/custom_actions/custom_actions_spec.rb
119:    cf = create(:list_wp_custom_field, multi_value: true)

features/projects/projects_index_spec.rb
984:    let!(:list_custom_field) { create(:list_project_custom_field, multi_value: true) }

features/work_packages/new/attributes_from_filter_spec.rb
80:        multi_value: true,

features/work_packages/table/switch_types_spec.rb
280:        multi_value: true,

features/projects/edit_settings_spec.rb
140:    let!(:list_custom_field) { create(:list_project_custom_field, name: 'List CF', multi_value: true) }
```

**Comment by Rincewind34** [comment]
Ok, I looked through all the resources and I think I know where to start (or at least what test cases to write). But I'll put this on the back burner for now and rather start with simpler frontend test cases that involve the other easier PRs I created :). Once I'm warming up with that, I'll be tackling this issue again.

**Comment by oliverguenther** [comment]
Closing this in favor of https://community.openproject.org/projects/openproject/work_packages/48745/activity
